src/Diagnostics/FxCop/Desktop.Analyzers/CSharp/Shared/SyntaxNodeHelper.cs

```diff
@@ -0,0 +1,262 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Linq;
+
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace Desktop.Analyzers.Common
+{
+    public sealed class CSharpSyntaxNodeHelper : SyntaxNodeHelper
+    {
+        private static CSharpSyntaxNodeHelper instance = new CSharpSyntaxNodeHelper();
+
+        public static CSharpSyntaxNodeHelper Default { get { return instance; } }
+
+        private CSharpSyntaxNodeHelper()
+        {} 
+
+        public override ITypeSymbol GetEnclosingTypeSymbol(SyntaxNode node, SemanticModel semanticModel)
+        {
+            if (node == null)
+            {
+                return null;
+            }
+
+            ClassDeclarationSyntax declaration = node.AncestorsAndSelf().OfType<ClassDeclarationSyntax>().FirstOrDefault();
```

Comment:
@genlu We should not traverse ancestors or descendants in any syntax node helpers, and any such syntax node helpers should be removed. For example, you shouldn't need the `GetEnclosingTypeSymbol(node)` helper here. Instead the correct way is to have the analyzer register:

1. A code block start action with a nested syntax node action
2. In the start action save off the owning member symbol
3. In the nested syntax action refer to the owning member's containing type.

Similarly you can get rid of other node helper methods that walk up or down the syntax tree.